images/encryption/encryption.go

```diff
@@ -0,0 +1,332 @@
+   limitations under the License.
+*/
+
+package encryption
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"strings"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images/encryption/blockcipher"
+	"github.com/containerd/containerd/images/encryption/config"
+	"github.com/containerd/containerd/images/encryption/keywrap"
+
+	"github.com/containerd/containerd/images/encryption/keywrap/jwe"
+	"github.com/containerd/containerd/images/encryption/keywrap/pgp"
+	"github.com/containerd/containerd/images/encryption/keywrap/pkcs7"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+// LayerInfo holds information about an image layer
+type LayerInfo struct {
+	// The Number of this layer in the sequence; starting at 0
+	Index      uint32
+	Descriptor ocispec.Descriptor
+}
+
+// LayerFilter holds criteria for which layer to select
+type LayerFilter struct {
+	// IDs of layers to touch; may be negative number to start from topmost layer
+	// empty array means 'all layers'
+	Layers []int32
+	// Platforms to touch; empty array means 'all platforms'
+	Platforms []ocispec.Platform
+}
+
+func init() {
+	keyWrappers = make(map[string]keywrap.KeyWrapper)
+	keyWrapperAnnotations = make(map[string]string)
+	RegisterKeyWrapper("pgp", pgp.NewKeyWrapper())
+	RegisterKeyWrapper("jwe", jwe.NewKeyWrapper())
+	RegisterKeyWrapper("pkcs7", pkcs7.NewKeyWrapper())
+}
+
+var keyWrappers map[string]keywrap.KeyWrapper
+var keyWrapperAnnotations map[string]string
+
+// RegisterKeyWrapper allows to register key wrappers by their encryption scheme
+func RegisterKeyWrapper(scheme string, iface keywrap.KeyWrapper) {
+	keyWrappers[scheme] = iface
+	keyWrapperAnnotations[iface.GetAnnotationID()] = scheme
+}
+
+// GetKeyWrapper looks up the encryptor interface given an encryption scheme (gpg, jwe)
+func GetKeyWrapper(scheme string) keywrap.KeyWrapper {
+	return keyWrappers[scheme]
+}
+
+// GetWrappedKeysMap returns a map of wrappedKeys as values in a
+// map with the encryption scheme(s) as the key(s)
+func GetWrappedKeysMap(desc ocispec.Descriptor) map[string]string {
+	wrappedKeysMap := make(map[string]string)
+
+	for annotationsID, scheme := range keyWrapperAnnotations {
+		if annotation, ok := desc.Annotations[annotationsID]; ok {
+			wrappedKeysMap[scheme] = annotation
+		}
+	}
+	return wrappedKeysMap
+}
+
+// EncryptLayer encrypts the layer by running one encryptor after the other
+func EncryptLayer(ec *config.EncryptConfig, encOrPlainLayerReader io.ReaderAt, desc ocispec.Descriptor) (content.ReaderDigester, map[string]string, error) {
+	var (
+		encLayerReader content.ReaderDigester
+		err            error
+		optsData       []byte
+	)
+
+	if ec == nil {
+		return nil, nil, errors.Wrapf(errdefs.ErrInvalidArgument, "EncryptConfig must not be nil")
+	}
+
+	symKey := make([]byte, 512/8)
+	_, err = io.ReadFull(rand.Reader, symKey)
```

Comment:
`rand.Read` uses `io.ReadFull` itself see https://golang.org/pkg/crypto/rand/#Read so you just need to use that.

Reply:
Calling miscreant's `GenerateKey()` now, which is good too since it checks for proper size...

https://github.com/stefanberger/containerd/commit/9f28224d4d2d3c5ee0317f019e22376e45c1fe9c